server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessor.java

```diff
@@ -272,36 +262,50 @@ private static void recordInAuditTrail(Mail mail, MailImpl copy, MailAddress ori
         void apply(Mail mail) throws Exception;
     }
 
-    public void processForwards(Mail mail) throws MessagingException {
+    public void processForwards(Mail mail) {
         if (rewriteSenderUponForward) {
+            LoopPrevention.RecordedRecipients recordedRecipients = LoopPrevention.RecordedRecipients.fromMail(mail);
             mail.getRecipients()
                 .stream()
-                .flatMap(Throwing.function(this::processForward))
+                .flatMap(Throwing.function(mailAddress -> processForward(mailAddress, recordedRecipients)))
                 .forEach(Throwing.consumer(decision -> decision.apply(mail)));
         }
     }
 
-    private Stream<ForwardDecision> processForward(MailAddress recipient) throws RecipientRewriteTableException {
-        ImmutableSet<Mapping> forwards = virtualTableStore.getStoredMappings(MappingSource.fromMailAddress(recipient))
-            .select(Mapping.Type.Forward)
-            .asStream()
-            .collect(ImmutableSet.toImmutableSet());
-
+    private Stream<ForwardDecision> processForward(MailAddress recipient, LoopPrevention.RecordedRecipients recordedRecipients) throws RecipientRewriteTableException {
+        ImmutableSet<Mapping> forwards = getForwards(recipient);
         if (forwards.isEmpty()) {
             return Stream.of();
         }
         Mapping localCopyMapping = Mapping.forward(recipient.asString());
         boolean localCopy = forwards.contains(localCopyMapping);
-        List<MailAddress> forwardedRecipients = forwards.stream()
+        List<MailAddress> forwardedRecipients = getForwardedMailAddressesWithLocalCopyExcluded(forwards, localCopyMapping);
+
+        Set<MailAddress> newRecipients = recordedRecipients.nonRecordedRecipients(ImmutableSet.copyOf(forwardedRecipients));
+        boolean shouldMailBeForwarded = !newRecipients.isEmpty();
+
+        if (localCopy && shouldMailBeForwarded) {
+            return Stream.of(ForwardDecision.sendACopy(mailetContext, recipient, forwardedRecipients, recordedRecipients, newRecipients));
+        }
+        if (!shouldMailBeForwarded) {
+            return Stream.of();
+        }
+        return Stream.of(ForwardDecision.removeRecipient(recipient),
+            ForwardDecision.sendACopy(mailetContext, recipient, forwardedRecipients, recordedRecipients, newRecipients));
```

Comment:
Too complicated to me. 

How about something more functionnal?

```
    Optional<Decision> localCopyDecision(ImmutableSet<Mapping> forwards, MailAddress recipient) {
        Mapping localCopyMapping = Mapping.forward(recipient.asString());
        if (forwards.contains(localCopyMapping) {
            return Optional.empty();
        }
        return Optional.of(ForwardDecision.removeRecipient(recipient));
    }

    Optional<Decision> forwardDecision(ImmutableSet<Mapping> forwards, MailAddress recipient) {
        Mapping localCopyMapping = Mapping.forward(recipient.asString());
        List<MailAddress> forwardedRecipients = getForwardedMailAddressesWithLocalCopyExcluded(forwards, localCopyMapping);
        if (forwardedRecipients.isEmpty()) {
            return Optional.empty();
        }
        return Optional.of(sendACopy(...));
    }

    
    Then
    
    return Stream.of(localCopyDecision(...),
        forwardDecision(...))
        .flatMap(Optional::stream);
```

Reply:
fixed